File: LeetCode/C#/572. Subtree of Another Tree.cs
public class Solution 
{
    public bool IsSubtree(TreeNode root, TreeNode subRoot)
	{
		if ((root == null && subRoot == null) && (root != null && subRoot == null))
			return true;
		else if (root == null && subRoot != null)
			return false;
		else // when both are not null
		{
			if (isSameTree(root, subRoot))  // Call our recursive function
				return true;
			else
				return IsSubtree(root.left, subRoot) || IsSubtree(root.right, subRoot); // Move down the next low Level
		} 
	}
	public bool isSameTree(TreeNode root, TreeNode subroot)
	{
		if (root == null && subroot == null)
			return true;
		else if (root != null && subroot != null && root.val == subroot.val)
			return isSameTree(root.left, subroot.left) && isSameTree(root.right, subroot.right);
		else    // if one of them is null
			return false;
	}
}

File: LeetCode/Python/572. Subtree of Another Tree.py
# type: ignore
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False
        
        if self.SameTree(root, subRoot):
            return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        
    def SameTree(self, root, subRoot):
        if not root and not subRoot:
            return True
        if root and subRoot and root.val == subRoot.val:
            return self.SameTree(root.left, subRoot.left) and self.SameTree(root.right, subRoot.right)
        return False


        